src/cmd/services/m3coordinator/ingest/ingest.go

```diff
@@ -0,0 +1,195 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ingest
+
+import (
+	"context"
+	"time"
+
+	"github.com/m3db/m3/src/cmd/services/m3coordinator/server/m3msg"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage"
+	"github.com/m3db/m3/src/query/ts"
+	"github.com/m3db/m3/src/x/serialize"
+	"github.com/m3db/m3metrics/metric/id"
+	"github.com/m3db/m3metrics/policy"
+	xerrors "github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/pool"
+	"github.com/m3db/m3x/retry"
+	xsync "github.com/m3db/m3x/sync"
+
+	"github.com/uber-go/tally"
+)
+
+// Options configures the ingester.
+type Options struct {
+	Appender          storage.Appender
+	Workers           xsync.PooledWorkerPool
+	PoolOptions       pool.ObjectPoolOptions
+	TagDecoderPool    serialize.TagDecoderPool
+	RetryOptions      retry.Options
+	InstrumentOptions instrument.Options
+}
+
+type ingestMetrics struct {
+	ingestError   tally.Counter
+	ingestSuccess tally.Counter
+}
+
+func newIngestMetrics(scope tally.Scope) ingestMetrics {
+	return ingestMetrics{
+		ingestError:   scope.Counter("ingest-error"),
+		ingestSuccess: scope.Counter("ingest-success"),
+	}
+}
+
+// Ingester ingests metrics with a worker pool.
+type Ingester struct {
+	workers xsync.PooledWorkerPool
+	p       pool.ObjectPool
+}
+
+// NewIngester creates an ingester.
+func NewIngester(
+	opts *Options,
+) *Ingester {
+	retrier := retry.NewRetrier(opts.RetryOptions)
+	m := newIngestMetrics(opts.InstrumentOptions.MetricsScope())
+	p := pool.NewObjectPool(opts.PoolOptions)
+	p.Init(
+		func() interface{} {
+			// NB: we don't need a pool for the tag decoder since the ops are
+			// pooled, but currently this is the only way to get tag decoder.
+			tagDecoder := opts.TagDecoderPool.Get()
+			op := ingestOp{
+				s:  opts.Appender,
+				r:  retrier,
+				it: serialize.NewMetricTagsIterator(tagDecoder, nil),
+				p:  p,
+				m:  m,
+			}
+			op.attemptFn = op.attempt
+			op.ingestFn = op.ingest
+			return &op
+		},
+	)
+	return &Ingester{
+		workers: opts.Workers,
+		p:       p,
+	}
+}
+
+// Ingest ingests a metric asynchronously with callback.
+func (i *Ingester) Ingest(
+	id []byte,
+	metricTimeNanos int64,
```

Comment:
Should we perhaps just make this `time.Time` considering it's converted to that later anyhow?  Less chance of getting the precision right if so.